pkg/cloudprovider/servicecontroller/servicecontroller.go

```diff
@@ -0,0 +1,468 @@
+		return err
+	} else if !exists {
+		return nil
+	}
+
+	if err := s.balancer.DeleteTCPLoadBalancer(s.loadBalancerName(service), s.zone.Region); err != nil {
+		return err
+	}
+	return nil
+}
+
+// ListKeys implements the interface required by DeltaFIFO to list the keys we
+// already know about.
+func (s *serviceCache) ListKeys() []string {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	keys := make([]string, 0, len(s.serviceMap))
+	for k := range s.serviceMap {
+		keys = append(keys, k)
+	}
+	return keys
+}
+
+func (s *serviceCache) get(serviceName string) (*cachedService, bool) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	service, ok := s.serviceMap[serviceName]
+	return service, ok
+}
+
+func (s *serviceCache) getOrCreate(serviceName string) *cachedService {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	service, ok := s.serviceMap[serviceName]
+	if !ok {
+		service = &cachedService{}
+		s.serviceMap[serviceName] = service
+	}
+	return service
+}
+
+func (s *serviceCache) set(serviceName string, service *cachedService) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.serviceMap[serviceName] = service
+}
+
+func (s *serviceCache) delete(serviceName string) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	delete(s.serviceMap, serviceName)
+}
+
+func needsUpdate(oldService *api.Service, newService *api.Service) bool {
+	if !oldService.Spec.CreateExternalLoadBalancer && !newService.Spec.CreateExternalLoadBalancer {
+		return false
+	}
+	if oldService.Spec.CreateExternalLoadBalancer != newService.Spec.CreateExternalLoadBalancer {
+		return true
+	}
+	if !portsEqual(oldService, newService) || oldService.Spec.SessionAffinity != newService.Spec.SessionAffinity {
+		return true
+	}
+	if len(oldService.Spec.PublicIPs) != len(newService.Spec.PublicIPs) {
+		return true
+	}
+	for i := range oldService.Spec.PublicIPs {
+		if oldService.Spec.PublicIPs[i] != newService.Spec.PublicIPs[i] {
+			return true
+		}
+	}
+	return false
+}
+
+// TODO: Use a shorter name that's less likely to be longer than cloud
+// providers' length limits.
+func (s *ServiceController) loadBalancerName(service *api.Service) string {
+	return s.clusterName + "-" + service.Namespace + "-" + service.Name
+}
+
+func getTCPPorts(service *api.Service) ([]int, error) {
+	ports := []int{}
+	for i := range service.Spec.Ports {
+		// TODO: Support UDP. Remove the check from the API validation package once
+		// it's supported.
+		sp := &service.Spec.Ports[i]
+		if sp.Protocol != api.ProtocolTCP {
+			return nil, fmt.Errorf("external load balancers for non TCP services are not currently supported.")
+		}
+		ports = append(ports, sp.Port)
+	}
+	return ports, nil
+}
+
+func portsEqual(x, y *api.Service) bool {
+	xPorts, err := getTCPPorts(x)
+	if err != nil {
+		return false
+	}
+	yPorts, err := getTCPPorts(y)
+	if err != nil {
+		return false
+	}
+	if len(xPorts) != len(yPorts) {
+		return false
+	}
+	// Use a map for comparison since port slices aren't necessarily sorted.
+	xPortMap := make(map[int]bool)
```

Comment:
The length check is already there, a couple lines above.

Comparing sorted lists would be neater, but it requires sorting them first, since I don't know of any restriction that a service's ports have to be listed in increasing order.